src/reporter/multi.rs

```
@@ -0,0 +1,131 @@
+use async_trait::async_trait;
+
+use crate::metadata::ReportMetadata;
+
+use super::Reporter;
+
+#[derive(Debug)]
+/// A reporter that reports profiling results to several destinations.
+///
+/// If one of the destinations errors, it will continue reporting to the other ones.
+pub struct MultiReporter {
+    reporters: Vec<Box<dyn Reporter + Send + Sync>>,
+}
+
+impl MultiReporter {
+    /// Create a new MultiReporter from a set of reporters
+    pub fn new(reporters: Vec<Box<dyn Reporter + Send + Sync>>) -> Self {
+        MultiReporter { reporters }
+    }
+}
+
+#[async_trait]
+impl Reporter for MultiReporter {
+    async fn report(
+        &self,
+        jfr: Vec<u8>,
+        metadata: &ReportMetadata,
+    ) -> Result<(), Box<dyn std::error::Error + Send>> {
+        let errs = futures::future::join_all(
+            self.reporters
+                .iter()
+                .map(|reporter| reporter.report(jfr.clone(), metadata)),
+        )
+        .await;
+        // return the first error
```

It seems bad to swallow errors past the first? Since they could be totally different destination types / failure reasons?

I don't think `Vec<Box<dyn std::error::Error + Send>>` is good because it is unpleasant to interact with compared to something directly implementing `std::error::Error`.

Ideally we would fold the errors somehow, but still preserves per-reporter failure attribution.

Perhaps its worth adding an optional `context()` method on the `Reporter` trait that we default to empty, implement for `S3Reporter`, etc?

But even without that we at least flatten down the errors in a way that maintains ordering if not direct context.